File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool IsValidBST(TreeNode root) {
        int prev = int.MinValue;
        Stack<TreeNode> stack = new();
        TreeNode curr = root;

        while (stack.Count > 0 || curr is not null) {
            while (curr is not null) {
                stack.Push(curr);
                curr = curr.left;
            }

            curr = stack.Pop();
            if (curr.val <= prev) {
                return false;
            }

            prev = curr.val;
            curr = curr.right;
        }

        return true;
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    sol = -Infinity;

    maxPathSum(root: TreeNode | null): number {
        const bruh = this.preorder(root);
        return this.sol;
    }

    preorder(node: TreeNode | null): number {
        if (!node) {
            return 0;
        }
        
        const left_max = Math.max(this.preorder(node.left), 0);
        const right_max = Math.max(this.preorder(node.right), 0);

        const curr_path = node.val + left_max + right_max;
        this.sol = Math.max(curr_path, this.sol);

        return node.val + Math.max(left_max, right_max);
    }
}
